### PR TITLE
Event forward in comp function services

### PR DESCRIPTION
**Pull request by zugao**
## Summary

* Adds event forwarding for all services under appcat. This applies to services that have their CRs managed by comp functions

## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Update tests.
- [x] Link this PR to related issues.

### CODE REVIEW

**Review by wejdross — APPROVED** [review]

**Review by Kidswiss — CHANGES_REQUESTED** [review]
Please make the PR name more descriptive, like `Event forwarding for external object storage`

File: pkg/comp-functions/runtime/function_mgr.go [hunk 1]
```diff
@@ -1064,3 +1069,45 @@ func (s *ServiceRuntime) addUsages() error {
 	}
 	return nil
 }
+
+func (s *ServiceRuntime) ForwardEvents() error {
+	claimRef := s.observedComposite.GetClaimReference()
```
Comment (Kidswiss) [inline pkg/comp-functions/runtime/function_mgr.go:1074]: We probably drop some events before the composite reference is available though.

Comment (zugao) [reply pkg/comp-functions/runtime/function_mgr.go:1074]: No they are not, and I don't even care. I had the same problem with desired, some times I don't have anything in there. Do you still want to go with desired?

**Comment by zugao** [comment]
> Please make the PR name more descriptive, like `Event forwarding for external object storage`

for all services :)

File: pkg/comp-functions/runtime/function_mgr.go (hunk 1 above)
Comment (Kidswiss) [reply pkg/comp-functions/runtime/function_mgr.go:1074]: I guess it's fine for now.

File: pkg/comp-functions/runtime/function_mgr.go [hunk 2]
```diff
@@ -1064,3 +1069,45 @@ func (s *ServiceRuntime) addUsages() error {
...
+				mrAnnotations = v.(map[string]any)
+			}
+			mrAnnotations[EventForwardAnnotation] = eventForwardValue
+			err := r.SetValue(p, mrAnnotations)
+			if err != nil {
+				return fmt.Errorf("cannot set event forward annotations for managed object %s: %w", r.GetName(), err)
+			}
+
+			// For non kube objects set directly into the managed resource
+		} else {
```
Comment (Kidswiss) [inline pkg/comp-functions/runtime/function_mgr.go:1098]: Shouldn't we also add the annotation to the outer kubeobject as well? Sometimes they might also fail to actually apply an object due to webhooks or other issues.

Comment (wejdross) [reply pkg/comp-functions/runtime/function_mgr.go:1098]: wouldn't that cause doubled entries in 99% of cases? outer object contains inner object's events and that would make readibility much worse

**Comment by Kidswiss** [comment]
> > Please make the PR name more descriptive, like `Event forwarding for external object storage`
> 
> for all services :) I updated the title

See, making it more descriptive helps :D

File: pkg/comp-functions/runtime/function_mgr.go (hunk 2 above)
Comment (Kidswiss) [reply pkg/comp-functions/runtime/function_mgr.go:1098]: Not necessarily, the outer object is more or less only concerned with applying the inner object to the cluster. For example if there's a webhook that blocks `.spec.important = false`, or if you try to apply a namespaced object without a namespace. That would only show up in the outer object. In such a case the object would not even reach the kubernetes api, and we'd have to describe the kubeobject to figure out why.

The events of the inner object are created by the controller that handles said object, only after its creation.

Comment (zugao) [reply pkg/comp-functions/runtime/function_mgr.go:1098]: I will add them to the kube Objects as well

**Review by TheBigLee — APPROVED** [review]

**Review by Kidswiss — APPROVED** [review]

**Pull request merged**